base/linalg/triangular.jl

```diff
@@ -995,15 +1048,97 @@ function logm{T<:Union{Float64,Complex{Float64}}}(A0::UpperTriangular{T})
          1.276404810806775e-001,
          2.060962623452836e-001,
          2.879093714241194e-001]
+
+    n = chksquare(A0)
+
+    A, m, s = invsquaring(A0,theta)
+
+    # Compute accurate diagonal of T - I
+    sqrt_diag!(A0, A, s)
+
+    # Compute the Gauss-Legendre quadrature formula
+    R = zeros(Float64, m, m)
+    for i = 1:m - 1
+        R[i,i+1] = i / sqrt((2 * i)^2 - 1)
+        R[i+1,i] = R[i,i+1]
+    end
+    x,V = eig(R)
+    w = Array(Float64, m)
+    for i = 1:m
+        x[i] = (x[i] + 1) / 2
+        w[i] = V[1,i]^2
+    end
+
+    # Compute the Padé approximation
+    Y = zeros(T, n, n)
+    for k = 1:m
+        Y = Y + w[k] * (A / (x[k] * A + I))
+    end
+
+    # Scale back
+    scale!(2^s, Y)
+
+    # Compute accurate diagonal and superdiagonal of log(T)
+    for k = 1:n-1
+        Ak = complex(A0[k,k])
+        Akp1 = complex(A0[k+1,k+1])
+        logAk = log(Ak)
+        logAkp1 = log(Akp1)
+        Y[k,k] = logAk
+        Y[k+1,k+1] = logAkp1
+        if Ak == Akp1
+            Y[k,k+1] = A0[k,k+1] / Ak
+        elseif 2 * abs(Ak) < abs(Akp1) || 2 * abs(Akp1) < abs(Ak)
+            Y[k,k+1] = A0[k,k+1] * (logAkp1 - logAk) / (Akp1 - Ak)
+        else
+            w = atanh((Akp1 - Ak)/(Akp1 + Ak) + im * pi * unw(logAkp1-logAk))
+            Y[k,k+1] = 2 * A0[k,k+1] * w / (Akp1 - Ak)
+        end
+    end
+    if isreal(Y)
+        return UpperTriangular(real(Y))
+    else
+        return UpperTriangular(Y)
+    end
+end
+logm(A::LowerTriangular) = logm(A.').'
+
+# Auxiliary functions for logm and matrix power
+
+function sqrt_diag!(A0::UpperTriangular, A::UpperTriangular, s)
+    n = chksquare(A0)
+    for i = 1:n
+        a = complex(A0[i,i])
+        if s == 0
+            r = a - 1
+        end
+        s0 = s
+        if angle(a) >= pi / 2
+            a = sqrt(a)
+            s0 = s - 1
+        end
+        z0 = a - 1
+        a = sqrt(a)
+        r = 1 + a
+        for j = 1:s0-1
+            a = sqrt(a)
+            r = r * (1 + a)
+        end
+        A[i,i] = z0 / r
```

Comment:
If `A` is a real matrix, won't this potentially throw an `InexactError` since `z0` is complex?  And if it doesn't, why do you need `a = complex(A0[i,i])` rather than `a = A0[i,i]`?